core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala

```diff
@@ -0,0 +1,546 @@
+  private val state = engineFactory.createEngine("scheduler")
+  private val stateLock = new ReentrantLock()
+  private val finishedDrivers = new mutable.ArrayBuffer[MesosClusterTaskState](retainedDrivers)
+  private val nextDriverNumber: AtomicLong = new AtomicLong(0)
+  private var frameworkId: String = null
+
+  // Stores all the launched and running drivers' states.
+  var launchedDrivers: LaunchedDrivers = _
+
+  // A queue that stores all the submitted drivers that hasn't been launched.
+  var queue: DriverQueue = _
+
+  // All supervised drivers that are waiting to retry after termination.
+  var superviseRetryList: SuperviseRetryList = _
+
+  private var masterInfo: Option[MasterInfo] = None
+
+  private def createDateFormat = new SimpleDateFormat("yyyyMMddHHmmss")  // For application IDs
+  private def newDriverId(submitDate: Date): String = {
+    "driver-%s-%04d".format(
+        createDateFormat.format(submitDate), nextDriverNumber.incrementAndGet())
+  }
+
+  def submitDriver(desc: MesosDriverDescription): CreateSubmissionResponse = {
+    val c = new CreateSubmissionResponse
+    val submitDate = new Date()
+    desc.submissionId = Some(newDriverId(submitDate))
+    desc.submissionDate = Some(submitDate)
+
+    stateLock.synchronized {
+      if (queue.isFull) {
+        c.success = false
+        c.message = "Already reached maximum submission size"
+      }
+      c.submissionId = desc.submissionId.get
+      queue.offer(desc)
+      c.success = true
+    }
+    c
+  }
+
+  def killDriver(submissionId: String): KillSubmissionResponse = {
+    val k = new KillSubmissionResponse
+    k.submissionId = submissionId
+    stateLock.synchronized {
+      // We look for the requested driver in the following places:
+      // 1. Check if submission is running or launched.
+      // 2. Check if it's still queued.
+      // 3. Check if it's in the retry list.
+      // 4. Check if it has already completed.
+      if (launchedDrivers.contains(submissionId)) {
+        val task = launchedDrivers.get(submissionId)
+        driver.killTask(task.taskId)
+        k.success = true
+        k.message = "Killing running driver"
+      } else if (queue.remove(submissionId)) {
+        k.success = true
+        k.message = "Removed driver while it's still pending"
+      } else if (superviseRetryList.remove(submissionId)) {
+        k.success = true
+        k.message = "Removed driver while it's retrying"
+      } else if (finishedDrivers.exists(s => s.submission.submissionId.equals(submissionId))) {
+        k.success = false
+        k.message = "Driver already terminated"
+      } else {
+        k.success = false
+        k.message = "Cannot find driver"
+      }
+    }
+    k
+  }
+
+  /**
+   * Recover scheduler state that is persisted.
+   * We still need to do task reconciliation to be up to date of the latest task states
+   * as it might have changed while the scheduler is failing over.
+   */
+  def recoverState: Unit = {
```

Comment:
This is not a getter without side-effects. The convention is to use `()` to denote such methods.